packages/tiny-graphql-query-compiler/src/index.ts

```diff
@@ -0,0 +1,140 @@
+export interface FieldSelection {
+  [key: string]: boolean | null | undefined | FieldCall | FieldSelection;
+}
+
+export interface BuilderOperation {
+  type: "query" | "subscription" | "mutation";
+  fields: FieldSelection;
+  name?: string;
+  directives?: string[];
+}
+
+const indent = (str: string, depth: number) => {
+  return str
+    .split("\n")
+    .map((line) => " ".repeat(depth) + line)
+    .join("\n");
+};
+
+const compileFieldSelection = (fields: FieldSelection, onlyPresentVariableValues = false): string => {
+  const result = Object.entries(fields)
+    .map(([field, value]) => {
+      if (typeof value === "boolean") {
+        return value ? field : false;
+      } else if (value instanceof FieldCall) {
+        let args = "";
+        let entries = Object.entries(value.args).filter(([_, value]) => value !== null && value !== undefined);
+
+        if (onlyPresentVariableValues) {
+          entries = entries.filter(([_, value]) => (value instanceof Variable ? value.present() : true));
+        }
+
+        const signatures = entries.map(([name, value]) => {
+          return `${name}: ${value instanceof Variable ? `$${value.name ?? name}` : JSON.stringify(value)}`;
+        });
+        if (signatures.length > 0) {
+          args = `(${signatures.join(", ")})`;
+        }
+
+        const subselection = value.subselection
+          ? `{ 
+${compileFieldSelection(value.subselection, onlyPresentVariableValues)} 
+}`
+          : "";
+
+        return `${field}${args} ${subselection}`;
+      } else {
+        return `${field} {
+${compileFieldSelection(value as FieldSelection, onlyPresentVariableValues)}
+}`;
+      }
+    })
+    .filter((value) => !!value)
+    .join("\n");
+
+  return indent(result, 2);
```

Comment:
nit: what do you think about avoiding the join + split + map + join by writing the above like:

```js
.filter((value) => !!value)
.map((value) => " ".repeat(2) + value)
.join("\n");
```